code/datums/beam.dm

```diff
@@ -26,53 +27,76 @@
 	icon_state = beam_icon_state
 	beam_type = btype
 
+
 /datum/beam/proc/Start()
 	Draw()
-	while(!finished && target && world.time<endtime && get_dist(origin,target)<max_distance && origin.z == target.z)
+	while(!finished && origin && target && world.time < endtime && get_dist(origin,target)<max_distance && origin.z == target.z)
 		if(origin.loc != origin_oldloc || target.loc != target_oldloc)
+			origin_oldloc = origin.loc //so we don't keep checking against their initial positions, leading to endless Reset()+Draw() calls
+			target_oldloc = target.loc
 			Reset()
 			Draw()
 		sleep(sleep_time)
+
 	qdel(src)
 
+
 /datum/beam/proc/End()
 	finished = 1
 
+
 /datum/beam/proc/Reset()
 	for(var/obj/effect/ebeam/B in elements)
 		qdel(B)
 
+
 /datum/beam/Destroy()
 	Reset()
 	target = null
 	origin = null
 	return ..()
 
+
 /datum/beam/proc/Draw()
-	var/Angle=round(Get_Angle(origin,target))
+	var/Angle = round(Get_Angle(origin,target))
 
 	var/matrix/rot_matrix = matrix()
 	rot_matrix.Turn(Angle)
 
-	var/DX=(32*target.x+target.pixel_x)-(32*origin.x+origin.pixel_x)
-	var/DY=(32*target.y+target.pixel_y)-(32*origin.y+origin.pixel_y)
-	var/N=0
-	var/length=round(sqrt((DX)**2+(DY)**2))
-	
-	for(N,N<length,N+=32)
-		var/obj/effect/ebeam/X= new beam_type(origin.loc)
-		X.owner=src
+	//Translation vector for origin and target
+	var/DX = (32*target.x+target.pixel_x)-(32*origin.x+origin.pixel_x)
+	var/DY = (32*target.y+target.pixel_y)-(32*origin.y+origin.pixel_y)
+	var/N = 0
+	var/length = round(sqrt((DX)**2+(DY)**2)) //hypotenuse of the triangle formed by target and origin's displacement
+
+	for(N in 0 to length-1 step 32)//-1 as we want < not <=, but we want the speed of X in Y to Z and step X
+		var/obj/effect/ebeam/X = new beam_type(origin.loc)
+		X.owner = src
 		elements |= X
+
+		//Assign icon, for main segments it's base_icon, for the end, it's icon+icon_state
+		//cropped by a transparent box of length-N pixel size
 		if(N+32>length)
-			var/icon/II=new(icon,icon_state)
+			var/icon/II = new(icon, icon_state)
 			II.DrawBox(null,1,(length-N),32,32)
-			X.icon=II
-		else X.icon=base_icon
+			X.icon = II
+		else
+			X.icon = base_icon
 		X.transform = rot_matrix
-		var/Pixel_x=round(sin(Angle)+32*sin(Angle)*(N+16)/32)
-		var/Pixel_y=round(cos(Angle)+32*cos(Angle)*(N+16)/32)
-		if(DX==0) Pixel_x=0
-		if(DY==0) Pixel_y=0
+
+		//Calculate pixel offsets (If necessary)
+		var/Pixel_x
+		var/Pixel_y
+		if(DX == 0)
+			Pixel_x = 0
+		else
+			Pixel_x = round(sin(Angle)+32*sin(Angle)*(N+16)/32)
```

Comment:
well, this is two sin() calls, related to an angle along X
the two cos() calls below relate to an angle along Y.

I used to have a mnemonic for this, something like "cos I'm walking up the stairs (Y axis going up)" but it's been 2 years since I needed that mnemonic so it's gone lol.

tl;dr it's just trigonometry.